spec.bs

```diff
@@ -507,22 +507,79 @@ To <dfn>validate and convert auction ad config</dfn> given an {{AuctionAdConfig}
-    |value|.
-  1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to
-    |signalsString|.
-1. If |config|["{{AuctionAdConfig/perBuyerTimeouts}}"] [=map/exists=], [=map/for each=] |key| →
-  |value| of |config|["{{AuctionAdConfig/perBuyerTimeouts}}"]:
-  1. If |key| equals to "*", then set |auctionConfig|'s [=auction config/all buyers timeout=]
-    to min(|value|, 500) milliseconds, and [=iteration/continue=].
-  1. Let |buyer| the result of [=parsing an origin=] with |key|. If |buyer| is failure,
-    [=exception/throw=] a {{TypeError}}.
-  1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer timeouts=][|buyer|] to
-    min(|value|, 500) milliseconds.
+1. If |config|["{{AuctionAdConfig/perBuyerSignals}}"] [=map/exists=]:
+  1. If |config|["{{AuctionAdConfig/perBuyerSignals}}"] is a {{Promise}}:
+    1. Set |auctionConfig|'s [=auction config/per buyer signals=] to
+      |config|["{{AuctionAdConfig/perBuyerSignals}}"].
+    1. Increment |auctionConfig|'s [=auction config/pending promise count=].
+    1. [=Upon fulfillment=] of |auctionConfig|'s [=auction config/per buyer signals=] with |result|:
+      1. Let |pendingException| be null.
+      1. If |result| is not a [=record=] mapping from {{USVString}} to {{any}}, set |pendingException| to
+        a {{TypeError}} and jump to the step labeled
+        <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
+      1. Set |auctionConfig|'s [=auction config/per buyer signals=] to a new [=ordered map=] whose
+        [=map/keys=] are [=origins=] and whose [=map/values=] are [=strings=].
+      1. [=map/For each=] |key| → |value| of |result|:
+        1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
+          set |pendingException| to a {{TypeError}} and jump to the step labeled
+          <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
+        1. Let |signalsString| be the result of [=serializing a JavaScript value to a JSON string=], given
+          |value|. If this [=exception/throws=] an exception |e|, set |pendingException| to |e| and jump
+          to the step labeled <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
+        1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to
+          |signalsString|.
+      1. <i id=validate-per-buyer-signals-finish>Finish</i>:
+        1. If |pendingException| is not null:
+          1. Set |auctionConfig|'s [=auction config/per buyer signals=] to failure.
+          1. Decrement |auctionConfig|'s [=auction config/pending promise count=].
+          1. [=exception/Throw=] |pendingException|.
+        1. Otherwise, decrement |auctionConfig|'s [=auction config/pending promise count=].
+    1. [=Upon rejection=] of |auctionConfig|'s [=auction config/per buyer signals=]:
+      1. Set |auctionConfig|'s [=auction config/per buyer signals=] to failure.

```

Comment:
Does this actually end up rejecting the outer runAdAuction promise ever?

Reply:
Yes, this is what Chrome's implementation does:
see impl here: https://chromium-review.googlesource.com/c/chromium/src/+/4114383/45/third_party/blink/renderer/modules/ad_auction/navigator_auction.cc#1075

see test InterestGroupBrowserTest.RunAdAuctionRejectPromisePerBuyerSignals

spec does it by line 717 below noting that line 538 set the [=auction config/per buyer signals=] to failure and returning failure which line 419 catches and then line 420 reject's the outer runAdAuction promise.